### PR TITLE
Feature/coupling of components

### PR DESCRIPTION
**Pull request by zlusch95**
This PR provides the computation of KPI Coupling Of Components.

### CODE REVIEW

File: src/entities/repositories/repository.service.ts [hunk 1]
```diff
@@ -9,6 +9,8 @@ import { CommitService } from '../commits/commit.service';
 import { ReleaseService } from '../releases/release.service';
 import { PullRequestService } from '../pullRequests/pullRequest.service';
 import {
+  // diffsPullrequestfilesLookup,
+  // issuesLabelsAssigneesLookup,
```
Comment (janniclas) [inline src/entities/repositories/repository.service.ts:13]: please remove

File: src/kpi/kpi.controller.ts [hunk 2]
```diff
@@ -60,6 +62,23 @@ export class KpiController {
...
+        owner: 'fraunhofer-iem',
+        repo: 'dodo-database-service',
+      },
+      100,
+      ['README.md'],
+      3,
+      2,
+      '2022-01-01',
+      '2022-02-15',
+    );
```
Comment (janniclas) [inline src/kpi/kpi.controller.ts:79]: please add query params to the method

File: src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    to: string = undefined,
+  ) {
+    const lookUpQuery = this.repoService.preAggregate(repoIdent, {
+      diffs: { prFiles: true, since: since, to: to },
+    });
+
+    const prFilesQuery = getPrFilesQuery(lookUpQuery, diffsLimit, fileFilter);
+
+    const prFiles = await prFilesQuery.exec();
+    console.log(prFiles);
```
Comment (janniclas) [inline src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts:37]: use the logger you created in line 10 !!!!

Comment (DieserFelix) [reply src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts:37]: Agreed

File: src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  // sorting map entries descending by occurences
+  const res = [...combinationMap.entries()].sort(
+    (a, b) => b[1]['occs'] - a[1]['occs'],
+  );
+
+  return res;
+
+  function increaseCounterAndAddPR(key: string, pr: string) {
+    const newAmount = combinationMap.get(key).occs + 1;
+    const newPRs = [...combinationMap.get(key).pullrequests, ...[pr]];
```
Comment (janniclas) [inline src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts:44]: you can just combinationMap.get(key).pullrequests.push(pr)
this should update the array stored in the map.
regardless of that, I'm not sure why you create a new string array by using '[pr]' and then unfold the created array with the spread operator ... 
this looks kind of wrong

File: src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  deleteOccurences(occs);
+
+  // sorting map entries descending by occurences
+  const res = [...combinationMap.entries()].sort(
+    (a, b) => b[1]['occs'] - a[1]['occs'],
+  );
+
+  return res;
+
+  function increaseCounterAndAddPR(key: string, pr: string) {
```
Comment (janniclas) [inline src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts:42]: pr: string is actually the pull request id I believe? 
please use clearer naming here

File: src/entities/repositories/repository.service.ts (hunk 1 above)
Comment (zlusch95) [reply src/entities/repositories/repository.service.ts]: removed

File: src/kpi/kpi.controller.ts (hunk 2 above)
Comment (zlusch95) [reply src/kpi/kpi.controller.ts:79]: added query params

File: src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts (hunk 3 above)
Comment (zlusch95) [reply src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts:37]: I am very sorry ... I forgot to delete this. It does not work to display it in that way with logger. Anyway, I got Postman :)

File: src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    const prAmount = prFiles.length;
+    console.log(prAmount);
```
[the remark this answers is not in the document]
Comment (zlusch95) [reply src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts:39]: Maybe we want to display all PRs which have been taken into account to have a ratio or something similar?
Like " in 9/45 PRs occurs this coupling"

File: src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts (hunk 4 above)
Comment (zlusch95) [reply src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts:44]: I remember the problem with this one; push() returns a number, I don't know why.
And therefore I found that way handy to concat to arrays, the old one which is stored already in the map with the new one, which is just a string of course, but regarding the push function we want to have it as an array.
I am open for a better way of handling this, but it works.

File: src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts (hunk 5 above)
Comment (zlusch95) [reply src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts:42]: Yes, you are right. I'll rename this.

File: src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts (hunk 4 above)
Comment (janniclas) [reply src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts]: it returns the new length of the array. so u actually update the array reference on which you push. so you don't need to reassign the array in the map. 
just push the objects in there and increase the number of occurrences. 

code example [here](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAtgTwLIEMAOMC8MwFMDuMqaAPNAE4CWYA5gDQwDeoArmFLuQFw4twBGnBmnIQeFajQDaAXQC+APgAUASgBQaxMQB0EXFCUAiDtEMNmINh24wATMNE8ph8LjMxj+EIfnq1oSFhgFnJyXHYANRQAGxZcLHhkdG0afSMTKEMVAEJ-cAgQaNxtaJAaJWDQ8Kgo2Nx1SrDImLjtVnZOGABqTABGGABufxCmmpbikQhtNBYIAAt0ubC3BvzC4tLyrWTUg2NcUxV1IA)

File: src/kpi/statistics/couplingOfComponents/lib/prFilesQuery.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { Aggregate, ObjectId } from 'mongoose';
+
+/**
+ * Extends preaggregated query @param lookUpQuery and
+ * aggregates all changed files for every PR. It exludes all
+ * files in @param fileFilter and applies the @param diffsLimit.
+ */
+export function getPrFilesQuery(
+  lookUpQuery: Aggregate<any[]>,
+  diffsLimit: number = 100,
```
[the remark this answers is not in the document]
Comment (janniclas) [reply src/kpi/statistics/couplingOfComponents/lib/prFilesQuery.ts:10]: correct

File: src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (janniclas) [reply src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts]: yeah but if I'm not mistaken we don't use the value, if we want to use it somehow we would need to return it from the api call.

[the remark this answers is not in the document]
Comment (zlusch95) [reply src/kpi/statistics/couplingOfComponents/couplingOfComponents.service.ts]: Yes, exactly, I return this now with the coupling.

**Pull request merged**